### PR TITLE
Add online scoring related classes for models

### PR DESCRIPTION
**Pull request by jq**
@deerzq @hectorgon @spencerde

### CODE REVIEW

File: core/src/main/java/com/airbnb/aerosolve/core/online/FeatureGen.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package com.airbnb.aerosolve.core.online;
+
+/*
+  use Float.MIN_VALUE as NULL for the float feature.
+ */
```
Comment (deerzq) [inline core/src/main/java/com/airbnb/aerosolve/core/online/FeatureGen.java]: add more comments describing what this class is doing

File: core/src/main/java/com/airbnb/aerosolve/core/online/FeatureMapping.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
+package com.airbnb.aerosolve.core.online;
+
+import lombok.Getter;
+
+import java.util.*;
+
+public class FeatureMapping {
```
Comment (deerzq) [inline core/src/main/java/com/airbnb/aerosolve/core/online/FeatureMapping.java]: more comments

File: core/src/main/java/com/airbnb/aerosolve/core/online/ModelType.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  private final String modelName;
+
+  @Getter
+  private final String path;
+
+  public ModelType(String modelName, String configName, String key) {
+    this.modelName = modelName;
+    this.configName = configName;
+    this.key = key;
+    path = "appraiser/" + modelName;
```
Comment (deerzq) [inline core/src/main/java/com/airbnb/aerosolve/core/online/ModelType.java]: "appraiser/"? change to something more generic

**Comment by deerzq** [comment]
It might worth writing a Readme file to describe how people can use this for online scoring, how this is different from the default AbstractModel, what extra functionality did you add.

File: core/src/main/java/com/airbnb/aerosolve/core/online/Features.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      Class c = field.getType();
+      if(c.equals(String.class)) {
+        features.add(new ScoringFeature(field.getName(), StringType));
+      } else if (c.equals(Double.class)) {
+        features.add(new ScoringFeature(field.getName(), FloatType));
+      } else if (c.equals(Boolean.class)) {
+        features.add(new ScoringFeature(field.getName(), BooleanType));
+      }
+    }
+    // Sort the non-amenity features alphabetically
```
Comment (spencerde) [inline core/src/main/java/com/airbnb/aerosolve/core/online/Features.java]: remove "non-amenity" here

File: core/src/main/java/com/airbnb/aerosolve/core/online/ModelType.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
+package com.airbnb.aerosolve.core.online;
+
+import lombok.Getter;
+
+public class ModelType {
+  @Getter
+  private final String configName;
+  @Getter
```
Comment (spencerde) [inline core/src/main/java/com/airbnb/aerosolve/core/online/ModelType.java]: nit add a new line for consistent spacing

**Comment by spencerde** [comment]
I agree that these classes will be a little tough for others to understand and use without more documentation and/or an example

**Comment by hectorgon** [comment]
Yeah without context I have no idea what this class is for

**Comment by jq** [comment]
@deerzq @spencerde  @hectorgon 
Add a unit test to show how to use these new classes. main goal is make it easy to turn features to example

File: core/src/main/java/com/airbnb/aerosolve/core/features/ExampleGen.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+/*
+  Generate Example from input features and defined featureFamily.
+  refer to ModelScorerTest.java as how to use ExampleGen
+ */
+public class ExampleGen {
+
+  public static Example toExample(Features features,
+                           List<StringFamily> stringFamilies,
+                           List<FloatFamily> floatFamilies) {
```
Comment (deerzq) [inline core/src/main/java/com/airbnb/aerosolve/core/features/ExampleGen.java]: need to consider `denseFeatures`

Comment (deerzq) [inline core/src/main/java/com/airbnb/aerosolve/core/features/ExampleGen.java]: 'Example' may consist of multiple `FeatureVector` and a context FeatureVector

**Comment by deerzq** [comment]
Please generalize this to take care of  `denseFeature`

File: core/src/main/java/com/airbnb/aerosolve/core/features/ExampleGen.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.*;
+
+/*
+  Generate Example from input features and defined featureFamily.
+  refer to ModelScorerTest.java as how to use ExampleGen
+ */
+public class ExampleGen {
+
+  public static Example toExample(Features features,
+                           List<StringFamily> stringFamilies,
```
Comment (deerzq) [inline core/src/main/java/com/airbnb/aerosolve/core/features/ExampleGen.java]: `toFeatureVector`?

**Comment by jq** [comment]
@deerzq  thanks for review, since this PR is big, let's add denseFeature in next PR, btw, it is easy to write DenseFamily extend FeatureFamily

**Comment by deerzq** [comment]
LGTM

**Comment by bhntr** [comment]
Despite the unit tests, I'm not sure I understand the usefulness of these classes? Why do we need to track the types in their own array?  Is there a reason `instanceof` won't work?

Is it helpful to assume that every feature family is homogenous (contains the same types)? Is this an assumption throughout aerosolve? If so and if it's useful, then shouldn't `FeatureFamily` be used throughout?

Also, it would be really helpful to have more javadoc explaining the classes and their properties.  I didn't find the comments very helpful to understanding the utility.

**Pull request merged**